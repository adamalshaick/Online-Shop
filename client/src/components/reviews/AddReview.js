import React, { Component } from "react";
import { connect } from "react-redux";
import { addReview } from "../../actions/reviewActions";
import ReviewForm from "../reviews/ReviewForm";
import { Button } from "../common/styles/Button";
import handleInputErrors from "../common/hoc/handleInputErrors";
import PropTypes from "prop-types";

export class AddReview extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showReviewInput: false,
      text: "",
      rate: ""
    };
  }

  onClickRev = () => {
    this.setState(prevState => ({
      showReviewInput: !prevState.showReviewInput
    }));
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = (id, name) => e => {
    e.preventDefault();
    const reviewData = {
      text: this.state.text,
      rate: this.state.rate,
      id: id,
      name: name
    };
    this.props.addReview(reviewData);
  };
  render() {
    const { errors, user } = this.props;
    return (
      <>
        <Button
          id="toggleBtn"
          className="btn btn-light btn-lg mt-2 mb-3"
          onClick={this.onClickRev}
        >
          Write a review
        </Button>
        <div>
          {this.state.showReviewInput ? (
            <ReviewForm
              id="reviewForm"
              onClickRev={this.onClickRev}
              text={this.state.text}
              rate={this.state.rate}
              errors={errors}
              onChange={this.onChange}
              onSubmit={this.onSubmit(user._id, user.name)}
            />
          ) : null}
        </div>
      </>
    );
  }
}

AddReview.propTypes = {
  addReview: PropTypes.func.isRequired,
  errors: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired
};

export default connect(
  null,
  { addReview }
)(handleInputErrors(AddReview));
