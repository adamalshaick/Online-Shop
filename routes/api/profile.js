const express = require("express");
const router = express.Router();
const passport = require("passport");

// Controllers
const profileController = require("../../controllers/profileController");

// @route GET api/profile
// @desc Get current users profile
// @access Private
router.get(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    profileController.getProfile(req, res);
  }
);

// @route GET api/profile/all
// @desc Get all profiles
// @access Public
router.get("/all", (req, res) => {
  profileController.getAllProfiles(req, res);
});

// @route GET api/profile/handle/:handle
// @desc Get profile by handle
// @access Public
router.get("/handle/:handle", (req, res) => {
  profileController.getProfileByHandle(req, res);
});

// @route GET api/profile/user/:user_id
// @desc Get profile by user ID
// @access Public

router.get("/user/:user_id", (req, res) => {
  profileController.getProfileById(req, res);
});

// @route POST api/profile
// @desc Create or edit user profile
// @access Private
router.post(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    profileController.createProfile(req, res);
  }
);

// @route GET api/profiles
// @desc Get profiles
// @access Public
router.get("/profiles", (req, res) => {
  profileController.getProfiles(req, res);
});

// @route   DELETE api/profile
// @desc    Delete user and profile
// @access  Private
router.delete(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    deleteProfile(req, res);
  }
);

module.exports = router;
