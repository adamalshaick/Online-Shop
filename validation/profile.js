const Validator = require("validator");
const isEmpty = require("./is-empty");

module.exports = function validateProfileInput(data, uploadError) {
  let errors = {};

  data.handle = !isEmpty(data.handle) ? data.handle : "";

  if (!Validator.isLength(data.handle, { min: 2, max: 40 })) {
    errors.handle = "Handle needs to be between 2 and 40 characters";
  }

  if (Validator.isEmpty(data.handle)) {
    errors.handle = "Username is required";
  }

  if (uploadError) {
    errors.file = uploadError;
  }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};
