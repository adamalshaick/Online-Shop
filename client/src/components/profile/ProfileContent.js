import React from "react";
import PropTypes from "prop-types";
import AddReview from "../reviews/AddReview";
import { connect } from "react-redux";
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";

export const ProfileContent = ({ user, auth }) => {
  return (
    <>
      <div className="text-center">
        <div className="mt-3">
          <TableRow>
            <TableCell>
              <strong>User's Name</strong>
            </TableCell>
            <TableCell align="center">{user.name}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>User's Location</strong>
            </TableCell>
            <TableCell align="center">{user.location}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>User's Bio</strong>
            </TableCell>
            <TableCell align="center">{user.bio}</TableCell>
          </TableRow>
        </div>
      </div>
      {user._id === auth.user.id ? null : <AddReview user={user} />}
    </>
  );
};

ProfileContent.propTypes = {
  user: PropTypes.object.isRequired,
  deleteAccount: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(ProfileContent);
