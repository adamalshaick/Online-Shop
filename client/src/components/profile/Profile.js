import React from "react";
import PropTypes from "prop-types";
import ProfileContent from "./ProfileContent";
import Reviews from "../reviews/Reviews";
import fetchUserById from "../common/hoc/fetchUserById";
import { SecondaryHeader } from "../common/styles/Header";

export const Profile = ({ user }) => {
  return (
    <div className="entry container mt-md-5">
      <div className="row">
        <section className="col-6">
          <SecondaryHeader className="text-center">
            Profile Info <hr />
          </SecondaryHeader>
          <ProfileContent user={user.user} />
        </section>
        <section className="col-6">
          <SecondaryHeader className="text-center">
            Reviews <hr />
          </SecondaryHeader>
          <Reviews user={user.user} />
        </section>
      </div>
    </div>
  );
};

Profile.propTypes = {
  user: PropTypes.object.isRequired
};

export default fetchUserById(Profile);
