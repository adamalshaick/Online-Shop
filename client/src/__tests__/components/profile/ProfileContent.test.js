import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { ProfileContent } from "../../../components/profile/ProfileContent";
import AddReview from "../../../components/reviews/AddReview";

const mockDeleteFn = jest.fn();

describe("render component", () => {
  const Props = {
    user: {},
    auth: { user: {} }
  };
  const wrapper = shallow(
    <ProfileContent {...Props} deleteAccount={mockDeleteFn} />
  );
  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("current user's profile", () => {
  const currentUserProps = {
    user: { _id: "matchId" },
    auth: { user: { id: "matchId" } }
  };
  const wrapper = shallow(
    <ProfileContent {...currentUserProps} deleteAccount={mockDeleteFn} />
  );

  it("shows delete profile button", () => {
    expect(wrapper.find("#delete").length).toEqual(1);
  });

  //   it("doesn't show add review button", () => {
  //     const component = wrapper.dive();
  //     expect(component.find(<AddReview />).length).toEqual(0);
  //   });

  //   it("calls mock delete profile function", () => {
  //     wrapper.find("#delete").simulate("click", { preventDefault() {} });
  //     expect(mockDeleteFn.mock.calls.length).toEqual(1);
  //   });
});

describe("other user's pofile", () => {
  const otherUserProps = {
    user: { _id: "matchId" },
    auth: { user: { id: "otherId" } }
  };
  const wrapper = shallow(
    <ProfileContent {...otherUserProps} deleteAccount={mockDeleteFn} />
  );

  it("doesn't show delete profile button", () => {
    expect(wrapper.find("#delete").length).toEqual(0);
  });

  //   it("shows add review button", () => {
  //     const component = wrapper.dive();
  //     expect(component.find(<AddReview />).length).toEqual(1);
  //   });
});
