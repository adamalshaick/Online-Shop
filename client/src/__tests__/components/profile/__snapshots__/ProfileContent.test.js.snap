// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<Fragment>
  <div
    className="text-center"
  >
    <div
      className="mt-3"
    >
      <WithStyles(TableRow)>
        <WithStyles(TableCell)>
          <strong>
            User's Name
          </strong>
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          align="center"
        />
      </WithStyles(TableRow)>
      <WithStyles(TableRow)>
        <WithStyles(TableCell)>
          <strong>
            User's Location
          </strong>
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          align="center"
        />
      </WithStyles(TableRow)>
      <WithStyles(TableRow)>
        <WithStyles(TableCell)>
          <strong>
            User's Bio
          </strong>
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          align="center"
        />
      </WithStyles(TableRow)>
    </div>
  </div>
  <div
    className="text-center mt-3"
  >
    <button
      className="btn btn-danger"
      id="delete"
      onClick={[Function]}
    >
      Delete Account
    </button>
  </div>
</Fragment>
`;
